go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+
+	numValues := int32(w.numBuffered)
+	nullCount := int32(pageStats.NullCount)
+	defLevelsByteLen := int32(defLevelsRLESize)
+	repLevelsByteLen := int32(repLevelsRLESize)
+
+	page := NewDataPageV2WithStats(memory.NewBufferBytes(combined), numValues, nullCount, numValues, w.encoding,
+		defLevelsByteLen, repLevelsByteLen, uncompressed, w.pager.HasCompressor(), pageStats)
+	if w.hasDict && !w.fallback {
+		w.totalCompressedBytes += int64(page.buf.Len()) // + sizeof pageheader
+		w.pages = append(w.pages, page)
+	} else {
+		w.totalCompressedBytes += int64(len(combined))
+		defer page.Release()
+		w.WriteDataPage(page)
+	}
+	return nil
+}
+
+func (w *columnWriter) FlushBufferedDataPages() {
+	if w.numBuffered > 0 {
+		w.AddDataPage()
+	}
+
+	for _, p := range w.pages {
+		defer p.Release()
+		w.WriteDataPage(p)
+	}
+	w.pages = w.pages[:0]
+	w.totalCompressedBytes = 0
+}
+
+func (w *columnWriter) writeLevels(numValues int64, defLevels, repLevels []int16) int64 {
+	toWrite := int64(0)
+	// if the field is required and non-repeated, no definition levels
+	if defLevels != nil && w.descr.MaxDefinitionLevel() > 0 {
+		for _, v := range defLevels {
+			if v == w.descr.MaxDefinitionLevel() {
+				toWrite++
+			}
+		}
+		w.WriteDefinitionLevels(defLevels[:numValues])
+	} else {
+		toWrite = numValues
+	}
+
+	if repLevels != nil && w.descr.MaxRepetitionLevel() > 0 {
+		// a row could include more than one value
+		//count the occasions where we start a new row
+		for _, v := range repLevels {
+			if v == 0 {
+				w.rowsWritten++
+			}
+		}
+
+		w.WriteRepetitionLevels(repLevels[:numValues])
+	} else {
+		// each value is exactly 1 row
+		w.rowsWritten += int(numValues)
+	}
+	return toWrite
+}
+
+func (w *columnWriter) writeLevelsSpaced(numLevels int64, defLevels, repLevels []int16) {
+	if w.descr.MaxDefinitionLevel() > 0 {
+		w.WriteDefinitionLevels(defLevels[:numLevels])
+	}
+
+	if w.descr.MaxRepetitionLevel() > 0 {
+		for _, v := range repLevels {
+			if v == 0 {
+				w.rowsWritten++
+			}
+		}
+		w.WriteRepetitionLevels(repLevels[:numLevels])
+	} else {
+		w.rowsWritten += int(numLevels)
+	}
+}
+
+func (w *columnWriter) WriteDictionaryPage() error {
+	dictEncoder := w.currentEncoder.(encoding.DictEncoder)
+	buffer := memory.NewResizableBuffer(w.mem)
+	buffer.Resize(dictEncoder.DictEncodedSize())
+	dictEncoder.WriteDict(buffer.Bytes())
+	defer buffer.Release()
+
+	page := NewDictionaryPage(buffer, int32(dictEncoder.NumEntries()), w.props.DictionaryPageEncoding())
+	written, err := w.pager.WriteDictionaryPage(page)
+	w.totalBytesWritten += written
+	return err
+}
+
+// this will always update the three otuput params
+// outValsToWrite, outSpacedValsToWrite, and NullCount. Additionally
+// it will update the validity bitmap if required (i.e. if at least one
+// level of nullable structs directly preced the leaf node)
+func (w *columnWriter) maybeCalculateValidityBits(defLevels []int16, batchSize int64, valuesToWrite, spacedValuesToWrite, nullCount *int64) {
```

Comment:
instead of taking the output values as args could we define a struct for them and return that?

Reply:
done